### PR TITLE
Review request for private event project

### PR DESCRIPTION
**Pull request by chrystalme**
## Private Event project
- A project for users to create invent that can be viewed by everyone but only signed-in users can attend.
- The project shows how associations work in rails. 
   - Belongs_to
   - has_many
   - has_many through
- scope 
- class_name

### CODE REVIEW

**Review by GzimAsani — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle: 

### Hello team, good job so far, but you have to do the following changes:
- [x] Your picture in readme file is broken, kindly fix that
- [ ] I cannot create new events I get the following error also delete this line `Events#new
Find me in app/views/events/new.html.erb`  Screenshot from 2021-03-05 21-49-04
- [x] Event model should belong to creator/user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

#### Controllers

##### EventsController

- [x] `index` action should: 
-  lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

### Please implement these changes and ask for another code review whenever you are ready

_**If you have any question related to the required changes feel free to text me on chat @GzimAsani**_

HAPPY CODING :computer: :desktop_computer:

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 🔴 ⚠️** 

## Hey @chrystalme & @ahmetbozaci, 👋🏽 
 **Thank you for the good job you got going on and making the requested changes**

## REQUIRED CHANGES

- [x] Add the linter files to your project. This will enable the github check to be done. You have 6 rubocop offenses locally.
- [ ] The event's show page should display a list of attendants of each event. As of now, all events belong to the signed-in user, even if they did not create the event, and have not opted to attend. You can choose to use a button that adds attendants of the event.
- [x] The index action of events controller should list both the past and upcoming events. Make the changes as requested in the previous review.

**Kindly make all the requested changes and submit another CRR. 👍🏽** 

## Cheers!! 🍹 & Happy Coding.

# Reach me on slack for a chitchat 🆘  **Roy Mukuye**

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required 🛑♻**

<hr>

Hello there, I must admit, great job working on this project so far and on the requested changes, it looks perfect. However, there are still a few more issues to make before moving on. :)

- [ ] Please consider making the upcoming and past events defined in the action index within the events controller. (As requested by previous reviewers)
- [ ] Kindly remove any related logic from your views and place it in the helpers as methods.

<hr>

Kindly make the required changes and ask for another CRR. :)
Good luck, happy coding, and keep up this great work. :)

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @chrystalme and @ahmetbozaci 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes  👇.
- [x] Kindly remove any related logic from your views and place it in the helpers as methods.
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by ahmetbozaci** [comment]
We move logic code from view as we can like everyone. 
We create method in helpers and we use partial in view.
I looked all TSE's project for how we do it. Everyone has logic in view.
TSE 's doesnt look code. Please check code.

File: app/views/events/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+<div class='container'>
+  <h1 class='title is-center'>Create Event</h1>
+  <%= form_with model: @created_event, url: event_create_path do |f| %>
+  <% if @created_event.errors.any? %>
+    <div class="notification is-danger">
+      <% @created_event.errors.full_messages.each do |message| %>
+        <div><%= message %></div>
+      <% end %>
+    </div>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/events/_form.html.erb:10]: there should be absolutely NO logic in views.

File: app/views/events/_show_events.html.erb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <h4> Past Events </h4>
+    <ul>
+      <% if @created_events.past_events.present? %>
+        <% @created_events.past_events.each do |event| %>
+        <li> <strong> <%= event.title %></strong> | <%= event.date %> </li>
+        <% end %>
+        </ul>
+      <% else %>  
+  <p> No past Events </p>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/events/_show_events.html.erb:32]: there should be absolutely NO logic in views.

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -1,17 +1,54 @@
...
+      <div class="navbar-item">
+
+        <div class="buttons">
+        <% if logged_in? %>
+           <h2 class='button is-static is-light'>Welcome  <%= current_user.name %></h2>
+          <%= link_to 'Log Out', logout_path, method: 'delete', class: 'button is-primary' %>
+        <% else %>
+          <%= link_to 'Sign Up', sign_up_path, class: 'button is-primary' %>
+          <%= link_to 'Sign In', sign_in_path, class: 'button is-light' %>
+        <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb:44]: there should be absolutely NO logic in views.

File: app/views/shared/_flash.html.erb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    <%= flash[:notice] %>
+  </div>
+<% end %>
+
+
+<% if flash[:alert] %>
+  <div class="notification is-danger">
+    <%= flash[:alert] %> 
+  </div>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_flash.html.erb:13]: there should be absolutely NO logic in views.

File: app/views/users/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  <h1 class='title is-center'>Sign Up</h1>
+  <%= form_with model: @user, url: sign_up_path do |f| %>
+  <% if @user.errors.any? %>
+    <div class="notification is-danger">
+      <% @user.errors.full_messages.each do |message| %>
+        <div><%= message %></div>
+      <% end %>
+    </div>
+  <% end %>
+  <div class="field">
```
Comment (acushlakoncept) [inline app/views/users/_form.html.erb:11]: there should be absolutely NO logic in views.

File: app/views/users/_show.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    </ul>
+  <% else %>
+    <p> You have no upcoming events. </p> 
+  <% end %>
+<% else %> 
+   <% Event.all.each do |created_event| %>
+        <h3 class='subtitle'> <%= created_event.title %> </h3>
+        <p> <%= created_event.description %> <span> <%= created_event.date %> </span></p>
+  <% end %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/users/_show.html.erb:28]: there should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @ahmetbozaci and @chrystalme, Great job! 👏  Both of you have done well on this project. Your UI looks good 👏👏👏

However, there are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.

- [x] I noticed there are lots of logic in your views, there should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

#### Tests

* [x] Your models should have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] Your authentication should have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [x] Your event managment features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))


### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by MahmoudBakr23 — APPROVED** [review]
# **Approved ✔✔**

<hr>

Hello Team :)
Good work you've done on this app and on the requested changes.
Thank you for the huge efforts :)
You may now merge and submit.
Good luck with the next and always happy coding 👨‍💻💻😊

**Pull request merged**